### PR TITLE
KNL-1333 - Adding support for hikaricp

### PR DESCRIPTION
**Pull request by jonespm**
This PR adds support for HikariCP. This could for sure use some testing, but it starts up without any warnings. There is some cleanup still but it's seems to be a great first (or in my case about sixth) pass

Some notes

* I had to turn off autocommit as that default to on and there's warnings if that is on. I think I'll just move that property into the XML since it's identical to Hikari so a slight re-factor.
* I believe that just setting registerMBeans has Hikari register them, previously on the tomcat driver we had to explicitly do that. I have not tested the mbeans
* It has compatibility with all of the old properties (url@ driverClassName@ etc), but I think to set the hikariCP properties we'll need to add them with defaults to the bean here.
* I only tested this with MySQL, same driver as the tomcat and dbcp ones

Java 7/8 note
They have two versions of the jar, one for Java 7 (which is currently the default in this patch) and one for Java 8. I'd anticipate making Java 8 the minimum requirement for Sakai 11, but for now this would (probably) not be compatible with java 8 without changing that dependency.

### CODE REVIEW

**Comment by jonespm** [comment]
To test this you can set a ${sakai.home}/sakai-configuration.xml as mentioned on 2.1.3 at https://confluence.sakaiproject.org/display/REL/More+Flexible+Sakai+Configuration and just use the name: javax.sql.hikaricp.BaseDataSource

**Comment by jonespm** [comment]
Also I believe it's breaking a test so I'll fix that with the other xml changes.

**Comment by jonespm** [comment]
I completely revised this today, exposing most of the settings via the bean that can be configured. I think a couple of them might be too low (they are the default) but everything seems to work. I think we're good to commit this after someone else starts it up. We'd probably want to change the default over to java 8 if java 8 support is completed, but that could be separate.

File: kernel/api/src/main/java/org/sakaiproject/hikaricp/jdbc/pool/SakaiBasicDataSource.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+	public synchronized void setRollbackOnBorrow(boolean value)
+	{
+		m_rollbackOnReturn = value;
+	}
+
+	public String getUrl() {
+        return url;
+    }
+
+    public void setUrl(String url) {
```
Comment (buckett) [inline kernel/api/src/main/java/org/sakaiproject/hikaricp/jdbc/pool/SakaiBasicDataSource.java]: Mix of indentation ( tabs/spaces ).

**Comment by jonespm** [comment]
I'll fix up the indentation, has anyone else tested this? Earle was saying he couldn't get it started. Sam was about to start it up.

**Comment by jonespm** [comment]
Rebased with indentation updated

**Comment by ottenhoff** [comment]
Okay looks good.

I did a basic test on an assignments page, and I notice that HikariCP returns less requests per second than Tomcat CP. But I assume this is because I'm not comparing the correct way:

> Q: I benchmarked Other Connection Pool X and HikariCP is coming out slower, what is happening?

> A: Make sure you are comparing apples-to-apples. While HikariCP strives to be high-performing it  also strives to be highly-reliabile. Many other pools' default settings are geared for performance over reliability. Additionally, unless you configure the underlying JDBC driver correctly, you are likely to miss out on performance. For example, other pools include a prepared statement cache of their own, while HikariCP relies on the caching ability of the underlying JDBC driver. Unless you enable this feature in the driver, you are losing performance. As an example, see the MySQL Configuration Tips.

**Comment by jonespm** [comment]
There are a lot of options on db-components.xml and I've exposed all of the ones that seem like anyone would want to set, including the dataSourceProperties. 

Are these good defaults? Maybe? Will the performance be better with Java 8? Does MySQL need a connectionTestQuery? (Previously validationQuery)

Locally I have validationQuery@javax.sql.BaseDataSource=select 1 from DUAL

This should probably be mapped so this old property = connectionTestQuery, so I'll fix that too.

		<property name="dataSourceProperties">
			<props>
				<prop key="cachePrepStmts">true</prop>
				<prop key="prepStmtCacheSize">250</prop>
				<prop key="prepStmtCacheSqlLimit">2048</prop>
				<prop key="useServerPrepStmts">true</prop>
			</props>
		</property>

**Comment by buckett** [comment]
Worked for me, I also set:
```
# So we can see our beans in JMX
registerMbeans@javax.sql.hikaricp.BaseDataSource=true
```
And could see the JMX component being setup and registered.

The Version 5.1 of Connector/J supports JDBC 4 I would expect that we can just use the connector's validation.

**Comment by ottenhoff** [comment]
Agree, that we should leave the validation query as the default so that we can take advantage of JDBC4 isValid calls.

The defaults look good to me. +1 for merging this in.

**Comment by jonespm** [comment]
Okay, merged this. It's 100% new additions and isn't enabled by default (at the moment) so no real possibility for any regression. We'll need to make a decision on Java 7 verses Java 8 still and also maybe a few other options, but can do that in another jira.

**Pull request merged**